README.md
# StringEncrypt Extension for Visual Studio Code

Encrypt and hide plain text strings & files contents within your source code in a secure & encrypted form with the help of a polymorphic engine, generating random-looking decryption code every time.

https://www.stringencrypt.com

Now available as a Visual Studio Code extension.

https://marketplace.visualstudio.com/items?itemName=PELock.stringencrypt

Simple & fast encryption — try it yourself!

## What it does?

It can turn your plain text strings or file contents into a random-looking encrypted mess. Take a look.

Before:

```js
var superSecretPhrase = "Easy string encryption for developers!";
```

After StringEncrypt:

```js
// encrypted with https://www.stringencrypt.com (v1.4.0) [JavaScript]
// superSecretPhrase = "Easy string encryption for developers!"
var superSecretPhrase = "\uBF59\uC4DE\uE963\uF128\uE48D\u0172\u0917\uF2DC" +
                        "\uFB41\u0A66\u0BCB\u1CF0\u1C95\u23BA\u2B5F\u3A44" +
                        "\u5929\u430E\u6953\u5378\u623D\u63A2\u74E7\u74AC" +
                        "\u8251\u8A37\u943C\u94C1\u9CE6\u830B\uACF0\uB3D5" +
                        "\uC23A\uCA5F\uCC44\uDA69\uBB0E\uEDD3";

for (var rOnpR = 0, FiuqR = 0; rOnpR < 38; rOnpR++)
{
        FiuqR = superSecretPhrase.charCodeAt(rOnpR);
        FiuqR -= rOnpR;
        FiuqR ^= 0xFFFF;
        FiuqR += 0x82A3;
        FiuqR = ((FiuqR << 5) | ( (FiuqR & 0xFFFF) >> 11)) & 0xFFFF;
        FiuqR += 0xE87C;
        FiuqR += rOnpR;
        FiuqR -= 0x7CB9;
        FiuqR = (((FiuqR & 0xFFFF) >> 7) | (FiuqR << 9)) & 0xFFFF;
        FiuqR ^= 0x4928;
        FiuqR += rOnpR;
        FiuqR --;
        FiuqR ^= 0xFC14;
        FiuqR -= 0x406C;
        FiuqR = (((FiuqR & 0xFFFF) >> 3) | (FiuqR << 13)) & 0xFFFF;
        superSecretPhrase = superSecretPhrase.substr(0, rOnpR) + String.fromCharCode(FiuqR & 0xFFFF) + superSecretPhrase.substr(rOnpR + 1);
}
```

## How to use it in Visual Studio Code?

StringEncrypt is available as an extension to the context menu activated with a right-mouse click on the active editor window.

### 1. Insert Encrypted String

1. Open the right-click menu in the active editor window
2. Click on the **Insert Encrypted String**
3. A dialog box will ask you to enter the **string label** e.g. _encryptedStringLabel_
4. A second dialog box will ask you to enter the **string** to be encrypted e.g. _"Hello, world!"_ (enter it without the quotes)

![Insert Encrypted String Example](https://www.stringencrypt.com/images/stringencrypt-insert-encrypted-string-example.gif)

> Note: In demo mode, label and an encrypted string length is only 6 characters max.

### 2. Encrypt Selected String

1. Select **any** text in the active editor window (we recommend entering it on a new line)
2. Open the right-click menu in the active editor window
3. Click on the **Encrypt Selected String**
4. A dialog box will ask you to enter the **string label** e.g. _encryptedStringLabel_

![Encrypt Selected String Example](https://www.stringencrypt.com/images/stringencrypt-encrypt-selected-string-example.gif)

> Note: This menu option is shown only if you select a string in your active editor window.

### 3. Insert Encrypted File

1. Open the right-click menu in the active editor window
2. Click on the **Insert Encrypted File**
3. A dialog box will ask you to select a file for encryption (it can be a binary file or a text file)
4. A second dialog box will ask you to enter the **string label** e.g. _encryptedStringLabel_

![Insert Encrypted File Example](https://www.stringencrypt.com/images/stringencrypt-insert-encrypted-file-example.gif)

> Note: This option is available only in the full version.

## Say hello to polymorphic encryption!

Forget about simple `xor` encryption! StringEncrypt comes with a unique encryption engine.

It's a **polymorphic encryption engine**, similar to the encryption methods used by the software protection solutions and advanced computer viruses.

### How it works?

Let me explain how the polymorphic encryption process works.

1. A random set of encryption commands is selected (`xor`, `addition`, `subtraction`, `bit rotations`, `bit shifts`, `logical negation` etc.).
2. A random set of helper `encryption keys` is generated.
3. Every byte of the input string is encrypted with every encryption command in the random set.
4. The decryption code in the selected programming language is generated with a reverse set of encryption commands.

More about polymorphic engines:

* How to build a polymorphic engine in C++ - https://www.pelock.com/articles/polymorphic-encryption-algorithms
* Polymorphic engine in 32-bit MASM assembler - https://github.com/PELock/Simple-Polymorphic-Engine-SPE32
* Poly Polymorphic Engine - https://www.pelock.com/products/poly-polymorphic-engine

### What does it mean?

The encrypted content is **different** every time you apply StringEncrypt encryption to it.

The algorithm is always **unique**, the encryption keys are always **randomly selected** and the decryption code is also **unique** for every time you use our encryption.

## Features

* Out of box support for `UNICODE` (WideChar type in `C/C++` languages), `UTF-8` (multibyte) & `ANSI` (single byte) strings encodings
* Configurable minimum & maximum number of encryption commands
* Different ways to store the encrypted string (as a `global` or `local` variable if the selected programming language supports it)
* Wide array of supported programming languages
* You can automate encryption process in your builds using our `WebAPI` interface

## Supported programming languages

StringEncrypt engine supports code generation for the following programming languages:

* [C & C++](https://www.stringencrypt.com/c-cpp-encryption/)
* [C# (C Sharp for .NET)](https://www.stringencrypt.com/c-sharp-encryption/)
* [Visual Basic .NET (VB.NET)](https://www.stringencrypt.com/visual-basic-net-vb-net-encryption/)
* [Delphi / Pascal](https://www.stringencrypt.com/delphi-pascal-encryption/)
* [Java](https://www.stringencrypt.com/java-encryption/)
* [JavaScript](https://www.stringencrypt.com/javascript-encryption/)
* [Python](https://www.stringencrypt.com/python-encryption/)
* [Ruby](https://www.stringencrypt.com/ruby-encryption/)
* [AutoIt](https://www.stringencrypt.com/autoit-encryption/)
* [Powershell](https://www.stringencrypt.com/powershell-encryption/)
* [Haskell](https://www.stringencrypt.com/haskell-encryption/)
* [MASM assembler](https://www.stringencrypt.com/masm-encryption/)
* [FASM assembler](https://www.stringencrypt.com/fasm-encryption/)

> Note: **Visual Studio Code** might not support some of the listed languages by default, you might install the extension first to support it (e.g. for Python), but you can always use our online interface to encrypt strings for it anyway.

## Available editions

StringEncrypt can be used:

* Directly on its website - https://www.stringencrypt.com/
* You can download standalone Windows client - https://www.stringencrypt.com/download/
* You can use it via WebAPI interface (from `PHP`) - https://www.stringencrypt.com/api/
* ...and **now** also as a Visual Studio Code extension - https://marketplace.visualstudio.com/items?itemName=PELock.stringencrypt

## Extension settings

You can fine-tune encryption settings in the extension **Settings** page.

## Free demo version limitations

The free demo version comes with some **limitations**.

| Feature                               | Demo version     | Full version |
|---------------------------------------|:----------------:|-------------:|
| String encryption                     | ✅              | ✅           |
| File encryption (text or binary file) | ❌              | ✅           |
| Max. label length (characters)        | `10`             | `64`         |
| Max. string length (characters)       | `10`             | `4096`       |
| Max. file length (bytes)              | —                | `4 MB`       |
| Min. number of encryption commands    | `3`              | `50`         |
| Max. number of encryption commands    | `3`              | `50`         |


## Purchase activation code

To remove the limitations and support our project and its development, you need to buy an activation code at:

https://www.stringencrypt.com/buy/

You can enter the activation code in the extension settings page. Each activation code has an assigned number of **usage credits**. You can use the software in full version as many times as you have usage credits on your account balance.

## How to get a free activation code?

You can get a **free activation code** (500 usage credits) if you can advertise StringEncrypt service with a link to the project site https://www.stringencrypt.com/ at:

* Programming forums
* Programming blogs
* Technical articles
* Twitter / Facebook / other social media site
* ...or any other website related to programming and development

Send me all the details at my [contact address](https://www.stringencrypt.com/) and if it's legit - **bam!**, you got yourself a free code :)

## Changelog

### [1.0.1] - 2020-07-26

- Fixes missing dependencies

### [1.0.0] - 2020-07-25

- Initial release of a StringEncrypt Extension for Visual Studio Code
- Support for string encryption
- Support for file encryption

CHANGELOG.md
# Change Log

## [1.0.1] - 2020-07-26

- Fixes missing dependencies


## [1.0.0] - 2020-07-25

- Initial release of a StringEncrypt Extension for Visual Studio Code
- Support for string encryption
- Support for file encryption


extension.js
"use strict";
const vscode = require('vscode');
const fs = require('fs');
const zlib = require('zlib');
const { StringEncrypt } = require('./stringencrypt');

const stringEncrypt = new StringEncrypt();

const labelPricing = "Check Full Version Pricing";
const labelIsItDown = "Check Server Status";
const labelBug = "Report a Bug";

// global activation status
var activationStatus;

function openWebPage(webPageUrl) {
	vscode.commands.executeCommand("vscode.open", vscode.Uri.parse(webPageUrl));
}

function openBuyPage() {
	openWebPage("https://www.stringencrypt.com/buy/");
}

function openContactPage() {
	openWebPage("https://www.stringencrypt.com/contact/");
}

function openIsItDown() {
	openWebPage("https://downforeveryoneorjustme.com/stringencrypt.com");
}

function connectionError() {
	vscode.window.showErrorMessage("Error while trying to connect to StringEncrypt.com website!", ...[labelIsItDown]).then(selection => { openIsItDown(); });
}

function getActiveEditor() {
	return vscode.window.activeTextEditor;
}

function getCurrentConfig() {
	return vscode.workspace.getConfiguration('stringencrypt');
}

function getProgrammingLanguage() {

	const editor = getActiveEditor();

	// is the language supported?
	if (!StringEncrypt.supportedLanguages.hasOwnProperty(editor.document.languageId)) {
		vscode.window.showErrorMessage('Language not supported by StringEncrypt. Please tell me about it!');
		openContactPage();
		return false;
	}

	return StringEncrypt.supportedLanguages[editor.document.languageId];
}

function formatBytes(bytes, decimals = 2) {
	if (bytes === 0) return '0 Bytes';

	const k = 1024;
	const dm = decimals < 0 ? 0 : decimals;
	const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];

	const i = Math.floor(Math.log(bytes) / Math.log(k));

	return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i];
}

async function refreshActivationStatus() {

	// get current configuration
	const config = getCurrentConfig();

	var result = stringEncrypt.checkActivationStatus(config.get("activationCode"),

		function (response) {

			// no response at all
			if (!response.data) {
				activationStatus = null;
				return;
			}

			activationStatus = response.data;

		}, function(error) {
			activationStatus = null;
			return;
		});

	return result;
}

function showActivationStatus() {

	// get current configuration
	const config = getCurrentConfig();

	stringEncrypt.checkActivationStatus(config.get("activationCode"),

		function (response) {

			// no response at all
			if (!response.data) {
				connectionError();
				return;
			}

			var info = response.data.demo ? "You are using a DEMO version of StringEncrypt.\n" : "You are using a FULL version of StringEncrypt.\r\n";
			info += `\r\n`;
			info += `Usage credits left - ${response.data.credits_left}\r\n`;
			info += `Usage credits total - ${response.data.credits_full}\r\n`;
			info += `\r\n`;
			info += `Max. label length - ${response.data.label_limit} characters\r\n`;
			info += `Max. string length - ${response.data.string_limit} characters\r\n`;
			info += `Max. bytes/file length - ${formatBytes(response.data.bytes_limit)}\r\n`;
			info += `\r\n`;
			info += `Min. number of encryption commands - ${response.data.cmd_min}\r\n`;
			info += `Max. number of encryption commands - ${response.data.cmd_max}\r\n`;

			vscode.window.showInformationMessage(info, { modal: true });

		}, connectionError);
}

function showGeneralInfo() {

	stringEncrypt.checkGeneralInfo(
		function (response) {

			// no response at all
			if (!response.data) {
				connectionError()
				return;
			}

			var info = `StringEncrypt engine version - ${response.data.engine_version}\r\n`;
			info += `\r\n`;
			info += `List of supported programming languages:\r\n`;
			info += `\r\n`;

			for (const [key, value] of Object.entries(response.data.supported_languages)) {

				const language = Object.entries(value);

				info += `${language[0][1]} (${language[0][0]})\r\n`;
			}

			vscode.window.showInformationMessage(info, { modal: true });

		}, connectionError);
}

/**
 * Encrypt text string
 *
 * @param {string} textLabel
 * @param {string} textString
 * @param {boolean} insertNewString
 */
async function encryptString(textLabel, textString, insertNewString) {

	// get active editor window
	const editor = getActiveEditor();

	// get current configuration
	const config = getCurrentConfig();

	// detect programming language in active editor window
	const langDetected = getProgrammingLanguage();

	if (!langDetected) return;

	var formData = {

		//
		// activation code, you can leave it empty for demo version, but keep in
		// mind that there are many limitations in demo version)
		//
		code : config.get("activationCode"),

		//
		// API command to execute
		//
		// "encrypt" - encrypt input string or file bytes, returns array of:
		//
		// $result["error"] - error code
		// $result["source"] - decryptor source code
		// $result["expired"] - activation code expiration flag (bool)
		// $result["credits_left"] - number of credits left
		//
		command: "encrypt",

		//
		// label name
		//
		// demo mode supports up to 6 chars only (64 in full version),
		// if you pass more than this number, service will return
		// ERROR_LENGTH_LABEL
		//
		label: textLabel,

		//
		// input string / raw bytes compression enabled, if you set it to
		// true, you need to compress input string / raw bytes eg.
		//
		// $compressed = @base64_encode(@gzcompress($string, 9)
		//
		// and after encryption you need to decompress encrypted data
		//
		// $decompressed = @gzuncompress(@base64_decode($source));
		//
		compression : false,
		//compression : true,

		//
		// input string in UTF-8 format
		//
		// demo mode supports up to 6 chars only, if you pass more
		// than that, service will return ERROR_LENGTH_STRING
		//
		string : textString,
		//string : @base64_encode(@gzcompress("Hello!", 9));

		//
		// raw data bytes to encrypt (you need to specify either
		// string or this value
		//
		// demo mode doesn't support this parameter and the service
		// will return ERROR_DEMO
		//
		//bytes : file_get_contents("my_file.txt");
		//bytes : file_get_contents("http://www.example.com/my_file.txt");
		//bytes : @base64_encode(@gzcompress(file_get_contents("my_file.txt"), 9));

		//
		// treat input string as a UNICODE string (ANSI otherwise)
		//
		unicode : config.get("useUnicode"),

		//
		// input string default locale (only those listed below
		// are supported currently)
		//
		lang_locale : "en_US.utf8",
		//lang_locale : "en_GB.utf8";
		//lang_locale : "de_DE.utf8";
		//lang_locale : "es_ES.utf8";
		//lang_locale : "fr_BE.utf8";
		//lang_locale : "fr_FR.utf8";
		//lang_locale : "pl_PL.utf8";

		//
		// how to encode new lines, available values:
		//
		// "lf" - Unix style
		// "crlf" - Windows style
		// "cr" - Mac style
		//
		new_lines : config.get("newLines").toLowerCase(),
		//new_lines : "lf";
		//new_lines : "crlf";
		//new_lines : "cr";

		//
		// destination ANSI string encoding (if unicode = false)
		//
		// only those listed below are supported
		//
		ansi_encoding : "WINDOWS-1250",
		//ansi_encoding:WINDOWS-1251";
		//ansi_encoding:WINDOWS-1252";
		//ansi_encoding:WINDOWS-1253";
		//ansi_encoding:WINDOWS-1254";
		//ansi_encoding:WINDOWS-1255";
		//ansi_encoding:WINDOWS-1256";
		//ansi_encoding:WINDOWS-1257";
		//ansi_encoding:WINDOWS-1258";
		//ansi_encoding:ISO-8859-1";
		//ansi_encoding:ISO-8859-2";
		//ansi_encoding:ISO-8859-3";
		//ansi_encoding:ISO-8859-9";
		//ansi_encoding:ISO-8859-10";
		//ansi_encoding:ISO-8859-14";
		//ansi_encoding:ISO-8859-15";
		//ansi_encoding:ISO-8859-16";

		//
		// output programming language
		//
		// only those listed below are supported, if you pass
		// other name, service will return ERROR_INVALID_LANG
		//
		lang : langDetected,
		//lang : "cpp",
		//lang : "csharp";
		//lang : "delphi";
		//lang : "java";
		//lang : "js";
		//lang : "python";
		//lang : "ruby";
		//lang : "autoit";
		//lang : "powershell";
		//lang : "haskell";
		//lang : "masm";
		//lang : "fasm";

		//
		// minimum number of encryption commands
		//
		// demo mode supports only up to 3 commands (50 in full version),
		// if you pass more than this number, service will return
		// ERROR_CMD_MIN
		//
		cmd_min : ( config.get("minEncryptionCommands") > activationStatus.cmd_max ) ? activationStatus.cmd_max : config.get("minEncryptionCommands"),
		//cmd_min : 1,

		//
		// maximum number of encryption commands
		//
		// demo mode supports only up to 3 commands (50 in full version),
		// if you pass more than this number, service will return
		// ERROR_CMD_MAX
		//
		cmd_max : ( config.get("maxEncryptionCommands") > activationStatus.cmd_max ) ? activationStatus.cmd_max : config.get("maxEncryptionCommands"),
		//cmd_max : 50,

		//
		// store encrypted string as a local variable (if supported
		// by the programming language), otherwise it's stored as
		// a global variable
		//
		local : config.get("localVariable"),
		//local : true,

		//
		// omit optional opening tags for the language e.g. <script></script>
		//
		include_tags : false,

		//
		// append optional example code
		//
		include_example : config.get("includeUsageExample")
	};

	stringEncrypt.webApiRequest(formData, function (response) {

		// no response at all
		if (!response.data) {
			connectionError()
			return;
		}

		// invalid success code
		if (response.data.error != StringEncrypt.errorCodes.ERROR_SUCCESS) {
			vscode.window.showErrorMessage(StringEncrypt.errorCodesLabelsMap.get(response.data.error));
			return;
		};

		// if there was selection -> replace it
		if (!editor.selection.isEmpty && !insertNewString) {

			editor.edit(builder => {
				builder.replace(editor.selection, response.data.source);
			});
		}
		else {
			editor.edit(builder => {
				builder.insert(editor.selection.active, response.data.source);
			});
		}

		if (response.data.credits_total == 0)
			vscode.window.showInformationMessage(`Encrypted string successfully generated (DEMO mode).`, ...[labelPricing]).then(selection => { openBuyPage(); });
		else
			vscode.window.showInformationMessage(`Encrypted string successfully generated (${response.data.credits_left} usage credits left).`);

		if (response.data.expired) {
			vscode.window.showWarningMessage('Your activation code has expired. Thank you for using StringEncrypt :)', ...[labelPricing]).then(selection => { openBuyPage(); });
		}

		//console.log(response);
	}, connectionError);
}

/**
 * Encrypt binary buffer
 *
 * @param {string} textLabel
 * @param {Buffer} bufferBytes
 */
async function encryptBuffer(textLabel, bufferBytes) {

	const editor = getActiveEditor();

	// get current configuration
	const config = getCurrentConfig();

	const langDetected = getProgrammingLanguage();

	if (!langDetected) return;

	// compress file contents using DEFLATE
	var fileBytesCompressed = zlib.deflateSync(bufferBytes);

	// compressed and BASE64 encoded
	var fileBytesCompressedAndEncoded = fileBytesCompressed.toString('base64');

	var formData = {

		//
		// activation code, you can leave it empty for demo version, but keep in
		// mind that there are many limitations in demo version)
		//
		code : config.get("activationCode"),

		//
		// API command to execute
		//
		// "encrypt" - encrypt input string or file bytes, returns array of:
		//
		// $result["error"] - error code
		// $result["source"] - decryptor source code
		// $result["expired"] - activation code expiration flag (bool)
		// $result["credits_left"] - number of credits left
		//
		command: "encrypt",

		//
		// label name
		//
		// demo mode supports up to 6 chars only (64 in full version),
		// if you pass more than this number, service will return
		// ERROR_LENGTH_LABEL
		//
		label: textLabel,

		//
		// input string / raw bytes compression enabled, if you set it to
		// true, you need to compress input string / raw bytes eg.
		//
		// $compressed = @base64_encode(@gzcompress($string, 9)
		//
		// and after encryption you need to decompress encrypted data
		//
		// $decompressed = @gzuncompress(@base64_decode($source));
		//
		compression : true,

		//
		// input string in UTF-8 format
		//
		// demo mode supports up to 6 chars only, if you pass more
		// than that, service will return ERROR_LENGTH_STRING
		//
		//string : textString,
		//string : @base64_encode(@gzcompress("Hello!", 9));

		//
		// raw data bytes to encrypt (you need to specify either
		// string or this value
		//
		// demo mode doesn't support this parameter and the service
		// will return ERROR_DEMO
		//
		//bytes : file_get_contents("my_file.txt");
		//bytes : file_get_contents("http://www.example.com/my_file.txt");
		//bytes : @base64_encode(@gzcompress(file_get_contents("my_file.txt"), 9));
		bytes : fileBytesCompressedAndEncoded,

		//
		// treat input string as a UNICODE string (ANSI otherwise)
		//
		unicode : false,

		//
		// input string default locale (only those listed below
		// are supported currently)
		//
		lang_locale : "en_US.utf8",
		//lang_locale : "en_GB.utf8";
		//lang_locale : "de_DE.utf8";
		//lang_locale : "es_ES.utf8";
		//lang_locale : "fr_BE.utf8";
		//lang_locale : "fr_FR.utf8";
		//lang_locale : "pl_PL.utf8";

		//
		// how to encode new lines, available values:
		//
		// "lf" - Unix style
		// "crlf" - Windows style
		// "cr" - Mac style
		//
		new_lines : config.get("newLines").toLowerCase(),
		//new_lines : "lf";
		//new_lines : "crlf";
		//new_lines : "cr";

		//
		// destination ANSI string encoding (if unicode = false)
		//
		// only those listed below are supported
		//
		ansi_encoding : "WINDOWS-1250",
		//ansi_encoding:WINDOWS-1251";
		//ansi_encoding:WINDOWS-1252";
		//ansi_encoding:WINDOWS-1253";
		//ansi_encoding:WINDOWS-1254";
		//ansi_encoding:WINDOWS-1255";
		//ansi_encoding:WINDOWS-1256";
		//ansi_encoding:WINDOWS-1257";
		//ansi_encoding:WINDOWS-1258";
		//ansi_encoding:ISO-8859-1";
		//ansi_encoding:ISO-8859-2";
		//ansi_encoding:ISO-8859-3";
		//ansi_encoding:ISO-8859-9";
		//ansi_encoding:ISO-8859-10";
		//ansi_encoding:ISO-8859-14";
		//ansi_encoding:ISO-8859-15";
		//ansi_encoding:ISO-8859-16";

		//
		// output programming language
		//
		// only those listed below are supported, if you pass
		// other name, service will return ERROR_INVALID_LANG
		//
		lang : langDetected,
		//lang : "cpp",
		//lang : "csharp";
		//lang : "delphi";
		//lang : "java";
		//lang : "js";
		//lang : "python";
		//lang : "ruby";
		//lang : "autoit";
		//lang : "powershell";
		//lang : "haskell";
		//lang : "masm";
		//lang : "fasm";

		//
		// minimum number of encryption commands
		//
		// demo mode supports only up to 3 commands (50 in full version),
		// if you pass more than this number, service will return
		// ERROR_CMD_MIN
		//
		cmd_min : config.get("minEncryptionCommands"),
		//cmd_min : 1,

		//
		// maximum number of encryption commands
		//
		// demo mode supports only up to 3 commands (50 in full version),
		// if you pass more than this number, service will return
		// ERROR_CMD_MAX
		//
		cmd_max : config.get("maxEncryptionCommands"),
		//cmd_max : 50,

		//
		// store encrypted string as a local variable (if supported
		// by the programming language), otherwise it's stored as
		// a global variable
		//
		local : config.get("localVariable"),
		//local : true,

		//
		// omit optional opening tags for the language e.g. <script></script>
		//
		include_tags : false,

		//
		// omit optional example code
		//
		include_example : config.get("includeUsageExample")
	};

	stringEncrypt.webApiRequest(formData, function (response) {

		// no response at all
		if (!response.data) {
			connectionError()
			return;
		}

		// invalid success code
		if (response.data.error != StringEncrypt.errorCodes.ERROR_SUCCESS) {
			vscode.window.showErrorMessage(StringEncrypt.errorCodesLabelsMap.get(response.data.error));
			return;
		};

		// first - decode from BASE64 format to raw buffer, then decompress it
		// catch any exceptions in decoding
		try {
			var data = zlib.inflateSync(Buffer.from(response.data.source, 'base64')).toString();

			editor.edit(builder => {
				builder.insert(editor.selection.active, data);
			});

			if (response.data.credits_total == 0)
				vscode.window.showInformationMessage(`Encrypted file successfully generated (DEMO mode).`, ...[labelPricing]).then(selection => { openBuyPage(); });
			else
				vscode.window.showInformationMessage(`Encrypted file successfully generated (${response.data.credits_left} usage credits left).`);

			if (response.data.expired) {
				vscode.window.showWarningMessage('Your activation code has expired. Thank you for using StringEncrypt :)', ...[labelPricing]).then(selection => { openBuyPage(); });
			}

		} catch (error) {
			vscode.window.showErrorMessage(`Error while decoding encrypted file!`, ...[labelBug]).then(selection => { openContactPage(); });
			//console.log(response);
		}

	}, connectionError);
}

async function queryInputString(placeHolderText, maxLength, messageDemo, messageFull) {

	var textString;

	while(true)
	{
		textString = await vscode.window.showInputBox({
			placeHolder: placeHolderText
		});

		// if text is empty - return
		if (!textString) return false;

		// if activationStatus is not available (something is wrong) enter
		if (!activationStatus) {
			vscode.window.showErrorMessage("Cannot connect to the StringEncrypt service to validate the license!");
			return false;
		}

		if (textString.length > maxLength) {

			if (activationStatus.demo) {
				vscode.window.showWarningMessage(messageDemo, ...[labelPricing]).then(selection => { openBuyPage(); });
			}
			else {
				vscode.window.showWarningMessage(messageFull);
			}
		}
		else break;
	}

	return textString;
}

async function stringEncryptReplace() {

	const editor = getActiveEditor();

	// refresh activation status using activation code in current configuration
	await refreshActivationStatus();

	if (!activationStatus) {
		connectionError();
		return;
	}

	var textLabel = await queryInputString("Enter label e.g. \"encryptedString\" for the encrypted string with StringEncrypt.",
		activationStatus.label_limit,
		`Label is too long! DEMO limit is ${activationStatus.label_limit} chars. Please enter it again.`,
		`Label is too long! Limit is ${activationStatus.label_limit} chars. Please enter it again.`
	);

	if (!textLabel) return;

	// selected text
	var textString = editor.document.getText(editor.selection);

	if (!textString) return;

	if (textString.length > activationStatus.string_limit) {
		vscode.window.showWarningMessage(activationStatus.demo ?
			`String is too long! DEMO limit is ${activationStatus.string_limit} chars. Please select a shorter string.` :
			`String is too long! Limit is ${activationStatus.string_limit} chars. Please select a shorter string.`
		);
		return;
	}

	return encryptString(textLabel, textString, false);
}

async function stringEncryptInsert() {

	// refresh activation status using activation code in current configuration
	await refreshActivationStatus();

	if (!activationStatus) {
		connectionError();
		return;
	}

	var textLabel = await queryInputString("Step 1. Enter label e.g. \"encryptedString\" for the encrypted string with StringEncrypt.",
		activationStatus.label_limit,
		`Label is too long! DEMO limit is ${activationStatus.label_limit} chars. Please enter it again.`,
		`Label is too long! Limit is ${activationStatus.label_limit} chars. Please enter it again.`
	);

	if (!textLabel) return;

	var textString = await queryInputString("Step 2. Enter string e.g. \"Hello, world!\" to be encrypted with StringEncrypt.",
		activationStatus.string_limit,
		`String is too long! DEMO limit is ${activationStatus.string_limit} chars. Please insert a shorter string.`,
		`String is too long! Limit is ${activationStatus.string_limit} chars. Please insert a shorter string.`
	);

	if (!textString) return;

	return encryptString(textLabel, textString, true);
}

async function stringEncryptInsertFile() {

	// refresh activation status using activation code in current configuration
	await refreshActivationStatus();

	if (!activationStatus) {
		connectionError();
		return;
	}

	// select input file (single file)
	var selectedFiles = await vscode.window.showOpenDialog({
		canSelectFiles : true,
		canSelectFolders : false,
		canSelectMany : false,
		title : "Step 1. Select any text or raw binary file for encryption with StringEncrypt"
	});

	if (!selectedFiles) return;

	if (activationStatus.demo) {
		vscode.window.showErrorMessage("This feature is not available in DEMO mode.", ...[labelPricing]).then(selection => { openBuyPage(); });
		return;
	}

	// try to read the input file
	var bufferFileBytes;

	try {
		bufferFileBytes = fs.readFileSync(selectedFiles[0].fsPath);
	} catch (error) {
		vscode.window.showErrorMessage("Cannot read the input file!");
		return;
	}

	// check file size (it cannot be empty)
	if (bufferFileBytes.length == 0) {
		vscode.window.showErrorMessage("Selected file is empty!");
		return;
	}

	if (bufferFileBytes.length > activationStatus.bytes_limit) {
		vscode.window.showWarningMessage(`File is too long (limit is ${formatBytes(activationStatus.bytes_limit)}. Please select a smaller file.`)
		return;
	}

	var textLabel = await queryInputString("Step 2. Enter label e.g. \"encryptedFile\" for the encrypted file contents with StringEncrypt.",
		activationStatus.label_limit,
		`Label is too long (DEMO limit is ${activationStatus.label_limit} chars. Please enter it again.`,
		`Label is too long (limit is ${activationStatus.label_limit} chars. Please enter it again.`
	);

	if (!textLabel) return;

	// send request to the WebAPI interface
	return encryptBuffer(textLabel, bufferFileBytes);
}

/**
 * @param {vscode.ExtensionContext} context
 */
async function activate(context) {

	let cmdConsole1 = vscode.commands.registerCommand('stringencrypt.checkActivationCode', function () {
		showActivationStatus();
	});

	let cmdConsole2 = vscode.commands.registerCommand('stringencrypt.getGeneralInfo', function () {
		showGeneralInfo();
	});

	let cmdMenu1 = vscode.commands.registerCommand('stringencrypt.stringEncryptInsert', function () {
		stringEncryptInsert();
	});

	let cmdMenu2 = vscode.commands.registerCommand('stringencrypt.stringEncryptReplace', function () {
		stringEncryptReplace();
	});

	let cmdMenu3 = vscode.commands.registerCommand('stringencrypt.stringEncryptInsertFile', function () {
		stringEncryptInsertFile();
	});

	// add console commands
	context.subscriptions.push(cmdConsole1);
	context.subscriptions.push(cmdConsole2);

	// add menus
	context.subscriptions.push(cmdMenu1);
	context.subscriptions.push(cmdMenu2);
	context.subscriptions.push(cmdMenu3);

	// detect any changed in the configuration
	vscode.workspace.onDidChangeConfiguration(async event => {
		if (event.affectsConfiguration('stringencrypt.activationCode')) {

			await refreshActivationStatus();

			if (activationStatus && activationStatus.demo == false) {
				vscode.window.showInformationMessage(`StringEncrypt is now activated. Usage credits left ${activationStatus.credits_left} out of ${activationStatus.credits_full} total.`);
			}
			else {
				vscode.window.showInformationMessage('StringEncrypt is now in DEMO mode.', ...[labelPricing]).then(selection => { openBuyPage(); });
			}
		}
	});

}

exports.activate = activate;

// this method is called when your extension is deactivated
function deactivate() {}

module.exports = {
	activate,
	deactivate
}
